### ISSUE TITLE
Intermittent test failures (`cases/external_python_buffer_example.cfg`, etc.)

### ISSUE THREAD

**Issue by hansenms** [gadgetron/gadgetron#1032]
The `cases/external_equivalent_example.cfg` integration test fails with some regularity. A repeat of the build pipeline usually fixes it.

This is the log snippet:

```
Writing logfile test/gadgetron.log to stdout:
12-09 23:20:22.789 INFO [main.cpp:76] Gadgetron 4.1.2 [a9cc375193d4ce8348d99d5ad8d49f18113d885b]
12-09 23:20:22.789 INFO [main.cpp:77] Running on port 9003
12-09 23:20:22.790 INFO [Server.cpp:25] Gadgetron home directory: "/usr/local"
12-09 23:20:22.790 INFO [Server.cpp:26] Gadgetron working directory: "/tmp/gadgetron/"
12-09 23:20:25.198 INFO [Server.cpp:38] Accepted connection from: ::ffff:127.0.0.1
12-09 23:20:25.201 INFO [ConfigConnection.cpp:131] Connection state: [CONFIG]
12-09 23:20:25.205 DEBUG [ConfigConnection.cpp:73] Reading config file: "/usr/local/share/gadgetron/config/external_equivalent_example.xml"
12-09 23:20:25.205 INFO [HeaderConnection.cpp:84] Connection state: [HEADER]
12-09 23:20:25.208 DEBUG [RESTStorageClient.cpp:301] Using storage address: http://localhost:9113
12-09 23:20:25.209 INFO [StreamConnection.cpp:75] Connection state: [STREAM]
12-09 23:20:25.241 DEBUG [Stream.cpp:58] Loading Gadget  of class SimpleReconGadget from gadgetron_mricore
12-09 23:20:25.242 DEBUG [Stream.cpp:58] Loading Gadget  of class NoiseAdjustGadget from gadgetron_mricore
12-09 23:20:25.242 DEBUG [NoiseAdjustGadget.cpp:195] Folder to store noise dependencies is /tmp/gadgetron
12-09 23:20:25.242 DEBUG [NoiseAdjustGadget.cpp:196] NoiseAdjustGadget::perform_noise_adjust_ is 1
12-09 23:20:25.242 DEBUG [NoiseAdjustGadget.cpp:197] NoiseAdjustGadget::pass_nonconformant_data_ is 1
12-09 23:20:25.242 DEBUG [NoiseAdjustGadget.cpp:198] receiver_noise_bandwidth_ is 0.793000
12-09 23:20:25.242 DEBUG [NoiseAdjustGadget.cpp:210] Measurement ID is 25135_461079_461084_37
12-09 23:20:25.242 DEBUG [Stream.cpp:58] Loading Gadget  of class AcquisitionAccumulateTriggerGadget from gadgetron_mricore
12-09 23:20:25.242 DEBUG [Stream.cpp:58] Loading Gadget  of class BucketToBufferGadget from gadgetron_mricore
12-09 23:20:25.243 ERROR [Core.cpp:18] [Connection Main Thread/Stream/NoiseAdjustGadget] ERROR: connect: Cannot assign requested address
12-09 23:20:25.246 DEBUG [Stream.cpp:58] Loading Gadget  of class ImageArraySplitGadget from gadgetron_mricore
12-09 23:20:25.247 DEBUG [Stream.cpp:58] Loading Gadget  of class ExtractGadget from gadgetron_mricore
12-09 23:20:25.247 DEBUG [Stream.cpp:58] Loading Gadget  of class RemoveROOversamplingGadget from gadgetron_mricore
12-09 23:20:25.247 DEBUG [RemoveROOversamplingGadget.cpp:43] RemoveROOversamplingGadget:omp_set_num_threads(1) ... 
12-09 23:20:25.247 DEBUG [Gadget.h:130] Shutting down Gadget ()
12-09 23:20:25.247 DEBUG [AcquisitionAccumulateTriggerGadget.cpp:129] Trigger (1) occurred, sending out 0 buckets
munmap_chunk(): invalid pointer
```

Seems to have problems assigning addresses, etc. Probably a timing issue with a previous test case.

**Comment by dchansen** [comment]
So I have been absolutely unable to reproduce this. 
We have been running the external_equivalent test continually for hours in the docker image on multiple machines and everything just works.
Our working hypothesis is that it is unique to Azure or at least to many core machines (current machine has 8 physical, 16 virtual cores). 

I did try to set up a 64 core machine on Azure, but that failed to convince customer service that I was not the world's smallest botnet.

**Comment by hansenms** [comment]
I can also confirm that running that single test for hours does not break it. It seems to be in the context of running other tests before. I think it has something to do with the test orchestration rather than the specific test. But when it fails it is one of the external tests. We see it frequently in the build pipeline and I have observed it locally as well in WSL2. It’s a pain to repro this.

**Comment by hansenms** [comment]
Updated the title of this. A better repro, actually seems to be to just run the `external_python_buffer_example.cfg` test over and over, revised from @dchansen repro in #1045:

```
while : ; do python3 run_gadgetron_test.py -p 1067 cases/external_python_buffer_example.cfg ; code=$?; echo $code; if [ $code -gt 0 ]; then break; fi ; done ; cat test/gadgetron.log
```

This often takes a looooong time to fail but seems to eventually.

**Comment by hansenms** [comment]
What we are getting is a connection refused:

```
01-27 22:30:20.887 DEBUG [RemoveROOversamplingGadget.cpp:43] RemoveROOversamplingGadget:omp_set_num_threads(1) ... 
01-27 22:30:20.887 INFO [Python.cpp:111] Started external Python module (pid: 2,957).
terminate called after throwing an instance of 'boost::wrapexcept<boost::system::system_error>'
  what():  connect: Connection refused
01-27 22:30:21.287 DEBUG [ext. 2957 gadgetron.examples.recon_buffers] Starting external Python module 'gadgetron.examples' in state: [ACTIVE]
01-27 22:30:21.287 DEBUG [ext. 2957 gadgetron.examples.recon_buffers] Connecting to parent on port 44011
01-27 22:30:21.289 CRITICAL [ext. 2957 gadgetron.examples.recon_buffers] [Errno 111] Connection refused
Traceback (most recent call last):
  File "/opt/conda/envs/gadgetron/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/envs/gadgetron/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/conda/envs/gadgetron/lib/python3.9/site-packages/gadgetron/__main__.py", line 55, in <module>
    raise e
  File "/opt/conda/envs/gadgetron/lib/python3.9/site-packages/gadgetron/__main__.py", line 52, in <module>
    main(sys.argv[1:])
  File "/opt/conda/envs/gadgetron/lib/python3.9/site-packages/gadgetron/__main__.py", line 45, in main
    with connection.Connection(socket.create_connection(address, 30)) as conn:
  File "/opt/conda/envs/gadgetron/lib/python3.9/socket.py", line 844, in create_connection
    raise err
  File "/opt/conda/envs/gadgetron/lib/python3.9/socket.py", line 832, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused
```

We don't get the whole log from the external process but the connection refused is pretty constant.

**Comment by kristofferknudsen** [comment]
I believe we do see the whole log from the external process. Specifically, the connection refused happens on [this](https://github.com/gadgetron/gadgetron-python/blob/57dfba04bc5b608a48a60527ee812200fde6ff50/gadgetron/__main__.py#L45) line, as the external process attempts to connect back to Gadgetron on the port specified. The only log messages before this point are present in the output, and the stack trace looks complete to me. 

For reference, [this](https://github.com/gadgetron/gadgetron/blob/b51c2c1a4647b25bb43cca2f52a94d32bc1ccd13/apps/gadgetron/connection/nodes/External.cpp#L69) is the Gadgetron side of the connection that fails:

```c++
    std::shared_ptr<ExternalChannel> External::open_connection(Config::Execute execute, const StreamContext &context) {

        tcp::endpoint endpoint(Info::tcp_protocol(), 0);
        auto acceptor = std::make_shared<tcp::acceptor>(io_service, endpoint); // *ACCEPTOR*

        auto port = acceptor->local_endpoint().port();
        boost::algorithm::to_lower(execute.type);

        GINFO_STREAM("Waiting for external module '" << execute.name << "' on port: " << port);

        auto child = std::make_shared<boost::process::child>(modules.at(execute.type)(execute, port, context)); // *CHILD*

        monitors.child = std::async(
                std::launch::async,
                [=](auto child, auto acceptor) { monitor_child(std::move(child), std::move(acceptor)); },
                child,
                acceptor
        ); // *MONITOR*

        auto socket = std::make_unique<tcp::socket>(io_service);
        auto future_socket = acceptor->async_accept(*socket, boost::asio::use_future); // *ACCEPT*

        io_service.run();
        future_socket.get();

        GINFO_STREAM("Connected to external module '" << execute.name << "' on port: " << port);

        auto stream = Gadgetron::Connection::stream_from_socket(std::move(socket));
        auto external_channel = std::make_shared<ExternalChannel>(
                std::move(stream),
                serialization,
                configuration
        );

        return external_channel;
    }
```

The flow is as follows:
Gadgetron needs to spawn an external child process, and establish an MRD stream to this child. 
It's pretty hard to know when the child will be ready to accept an MRD stream connection, so the Gadgetron parent will listen on a random free port, communicate this port to the child, who then connects back to parent when ready. Everything then proceeds when the connection is up. Nobody has to poll, or guess, or try repeatedly.

Once the acceptor is initialized, the OS machinery is in place to accept TCP connection on the chosen port. We then go on to spawn the external process, and start a thread to monitor the child (so we can stop waiting and fail out gracefully if the child exits before it can make the connection).

Gadgetron will then wait for the child to connect back - in this case, that connection is refused somehow. 

I'm low on ideas as to what could cause this connection to be refused. The acceptor is in place before we spawn the child - a sleep before the accept makes no difference when I test it, and I never see the issue when running locally.

**Comment by dchansen** [comment]
@hansenms I still can't reproduce this. Any chance you could give us the full log of a failing output? Including the tests that were not failing before.

**Comment by hansenms** [comment]
Closing this based on @kristofferknudsen's fix in #1049
